[master] Manage SSH keys and certificates

lkubb opened the pull request:
### What does this PR do?
Adds `ssh_pki` execution and state modules that apply the principle (and rough code outline) of the `x509` modules to SSH keys and certificates. This allows to statefully manage ssh keys and certificates and create an OpenSSH CA managed by Salt.

Note: Based on https://github.com/saltstack/salt/pull/65654.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/65197

### Merge requirements satisfied?
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### Notes
I'm not sure if adding a new module fits the current zeitgeist of moving modules out into extensions, but tend to think being able to manage SSH keys (and certificates) should be part of Salt core.

In a lucky coincidence, the development of these modules resulted in the discovery of https://github.com/advisories/GHSA-cf7p-gm2m-833m, which means Salt now ships with the required `cryptography` version for them by default. :)

lkubb committed:
Add ssh_pki modules

lkubb committed:
Revert to using `state.single` for `file.managed` calls

The underlying issue has been fixed in 3006.3

lkubb committed:
Cleanup docs

lkubb committed:
Misc cleanup/minor fixes

* do not try to load signing private key when different backend is in
  use and ca_server (which can be passed, but is not required) is None
* ensure `read_certificate` returns dates in UTC
* update tests certificate after cryptography encoding fix
* move safe_atomic_write to utils (incl. x509_v2)

lkubb committed:
Remove `days_remaining` and `days_valid` compatibility

This was untested and added unnecessary complexity

lkubb committed:
Add changelog

lkubb committed:
Fix unit test - `days_valid` was not untested after all

OrangeDog commented:
This feels like it should be sharing code with `x509_v2` and `tls` modules, as they all do essentially the same thing but with different default attributes. It would also avoid unexpected behaviour divergence.

lkubb commented:
@OrangeDog Although my initial expectation was the same, X509 v3 certificates are much more complicated than OpenSSH ones (and even for analogues, the `cryptography` APIs differ in some details). There might be some sections that can be abstracted somewhat reasonably regarding remote issuance and file handling, will investigate that further.

For the initial PR, I shied away from refactoring the `x509_v2` module to first gauge sentiment with a MVP. Also notice the modules in here are a bit more flexible than their `x509_v2` counterparts  - they allow signing policies to define overridable defaults + allowed values and the state module can optionally use a different execution module as the backend (there is an implementation for Vault e.g.). Implementing the latter for X509 is a lot of work.

I cannot speak for `tls`, but from what I groked it could be rewritten on top of `x509_v2` (or deprecated) since it's just a local CA implementation with defaults (what you would do with states otherwise).

OrangeDog commented:
> Implementing the latter for X509 is a lot of work

But you don't have to do that. If you've already implemented the flexibility, just call it from x509 without using those bits.

lkubb commented:
Sure, was trying to point out a few of the many subtle differences that carry the potential for the abstraction to end up as essentially two separate code paths in one ugly bowl of spaghetti.

I will investigate possible abstractions + do some low-hanging fruits as soon as I have the capacity, more complicated ones in case there is actual interest for this to be merged.

lkubb committed:
Refactor certificate_managed changes checks

to be able to use them in a wrapper module

lkubb committed:
Add x509_v2 SSH wrapper

Also fixes an issue where ``encoding`` could not be passed to
``file.managed`` states.

lkubb committed:
Refactor ssh_pki.certificate_managed changes checks

to be able to use them in a wrapper module

lkubb committed:
Add `ssh_pki` SSH wrapper module

lkubb committed:
Remove datetime.utcnow usage

lkubb committed:
Make certificate_managed wrapper overwrite unreadable private keys

lkubb committed:
Ensure symlinks are warned about and not followed

lkubb committed:
Handle existing invalid files better

lkubb committed:
Small fixes

lkubb committed:
Skip tests if host lacks the correct cryptography version

lkubb committed:
Remove _check_ret calls since they're not needed anymore

after #64542 was merged

lkubb committed:
Skip new tests on cryptography <3.1

`x509_v2` requires 3.1 because of PKCS7 import
This is only an issue on Debian 10 and Ubuntu 20.04.

lkubb committed:
Add changelog

lkubb commented:
@dwoz Is managing SSH keys and certificates a functionality the team would accept into core in 3008? (Asking because of the new policy regarding module additions).

Note: This branch is currently outdated, but my current working branch is based on https://github.com/saltstack/salt/pull/65654, so I will wait for now.

lkubb committed:
Fixes/corrections after rebase

lkubb committed:
Try to fix FIPS runs

lkubb committed:
fix(ssh_pki.public_key_managed): Don't fail in test mode

lkubb committed:
Run pre-commit after rebase

twangboy reviewed (changes requested):
A few docs changes

lkubb commented on salt/states/x509_v2.py:
```diff
@@ -1527,45 +1487,79 @@ def private_key_managed(
+def private_key_managed_ssh(name, result, comment, changes, tempfile=None, **kwargs):
+    """
+    Helper for the SSH wrapper module to report the correct return and
+    perform a ``file.managed`` sub-state run.
```
https://github.com/saltstack/salt/pull/64708#discussion_r2016760134 is valid here as well.

lkubb commented on salt/utils/x509.py:
```diff
@@ -477,6 +477,287 @@ def build_crl(
+):
+    """
+    Check if the on-disk certificate needs to be updated.
+    Extracted from the state module to be able to use this in the SSH wrapper.
```
This is part of https://github.com/saltstack/salt/pull/65654, which this PR is based on. Should I modify it there? Note that these functions should be considered module-internal ones since they are highly implementation-specific.

In this case, the (many) parameters are already documented on the modules that call this function.

lkubb commented on salt/utils/x509.py:
```diff
@@ -477,6 +477,287 @@ def build_crl(
+def compare_exts(current, builder):
+    """
+    Compare an extensions on an existing entity (certificate, CRL, CSR) to those
+    on a builder instance to check for changes.
```
Same as https://github.com/saltstack/salt/pull/64708#discussion_r2016869414

lkubb commented on salt/utils/x509.py:
```diff
@@ -477,6 +477,287 @@ def build_crl(
+
+def compare_ca_chain(current, new):
+    """
+    Compare two lists of certificates for equality.
```
Same as https://github.com/saltstack/salt/pull/64708#discussion_r2016869414

lkubb committed:
More docstrings, adjust versionadded

lkubb committed:
Rename integration tests

in CI, pytest complained about an `import file mismatch` because these
files have the same basename as the x509_v2 tests.

Output was:

```
ERROR collecting tests/pytests/integration/ssh/ssh_pki/test_certificate_managed_wrapper.py
import file mismatch:
imported module 'test_certificate_managed_wrapper' has this __file__ attribute:
  /__w/salt/salt/tests/pytests/integration/ssh/x509_v2/test_certificate_managed_wrapper.py
which is not the same as the test file we want to collect:
  /__w/salt/salt/tests/pytests/integration/ssh/ssh_pki/test_certificate_managed_wrapper.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```

lkubb committed:
Adjustments after Salt-SSH retcode fixes

lkubb committed:
Skip tests when missing requirements

... and remove unnecessary fixture requests.

lkubb committed:
Fix system python check after rebase
